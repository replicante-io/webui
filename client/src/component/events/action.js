'use strict';
//@flow


export type Event = {
  event: string,
  payload: any,
  timestamp: string,
};


/** Name actions for type checking. */
export const EVENTS_FETCH = 'EVENTS_FETCH';
export const EVENTS_STORE = 'EVENTS_STORE';


/** Type enum of all possible fetch actions. */
export type EventsFetchAction = {
  +type: typeof EVENTS_FETCH,
};

export type EventsStoreAction = {
  +type: typeof EVENTS_STORE,
  +events: Array<Event>,
};

export type EventsAction =
  EventsFetchAction |
  EventsStoreAction |
  {type: 'FLOW_CATCH_ALL'};


export function fetchEvents(): EventsFetchAction {
  return {
    type: EVENTS_FETCH
  };
}


export function storeEvents(events: Array<Event>): EventsStoreAction {
  return {
    type: EVENTS_STORE,
    events: events,
  };
}
