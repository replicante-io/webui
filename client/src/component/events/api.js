'use strict';
//@flow

import type { Event } from './action';


export function fetchEvents(): Promise<Array<Event>> {
  let url = '/api/events';
  return fetch(url).then((response) => {
    return response.json().then((body) => {
      if (!response.ok) {
        throw Error('Fetch error: ' + body.error);
      }
      return body;
    });
  });
}
