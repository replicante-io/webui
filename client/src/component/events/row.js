'use strict';
//@flow

import React from 'react';
import ReactJson from 'react-json-view';
import { Link } from 'react-router-dom';

import type { Event } from './action';


class EventRow extends React.Component<Event> {
  render() {
    return (
      <div className="row event-item">
        <div className="col timestamp">
          {this.props.timestamp}
        </div>
        <div className="col event">
          {this.props.event}
        </div>
        <div className="col-6 payload codeblock">
          <ReactJson
              collapsed={0}
              name={false}
              src={this.props.payload}
          />
        </div>
      </div>
    );
  }
}
export default EventRow;
