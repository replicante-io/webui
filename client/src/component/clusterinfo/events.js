'use strict';
//@flow

import React from 'react';
import { connect } from 'react-redux';

import { fetchEvents } from './action';
import { NoDataIcon } from '../icons';
import EventRow from '../events/row';

import type { Element } from 'react';
import type { ClusterDiscovery } from './action';
import type { ClusterInfoStore } from './store';
import type { Event } from '../events/action';


type Props = {
  events: ?Array<Event>,
  dispatch: any,
  id: string,
};
export class InnerEvents extends React.Component<Props> {
  componentDidMount() {
    if (!this.props.events) {
      this.props.dispatch(fetchEvents(this.props.id));
    }
  }

  render() {
    if (this.props.events) {
      // $FlowFixMe: flow can't cope with the return type of the map.
      return this.props.events.map((event, idx) => {
        return <EventRow key={idx} {...event} />;
      });
    }
    return (
      <div className="text-center">
        <NoDataIcon />
        No information available. <br />
        If nothing shows up the cluster may not exist.
      </div>
    );
  }
}

/** Map the redux state to properties passed to Button. */
type PartialState = {
  clusterinfo: ClusterInfoStore,
}
export function mapStateToProps(state: PartialState, props: {match: any}) {
  const id = props.match.params.name;
  return {
    events: state.clusterinfo.events[id],
    id: id,
  };
}


const Events = connect(mapStateToProps, null)(InnerEvents);
export default Events;
