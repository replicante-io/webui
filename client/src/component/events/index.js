'use strict';
//@flow

import React from 'react';
import { connect } from 'react-redux';

import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import fasChartArea from '@fortawesome/fontawesome-free-solid/faChartArea';

import { fetchEvents } from './action';
import EventRow from './row';
import type { Event } from './action';
import type { EventsStore } from './store';


// Re-export actions and store values.
export { defaultState, reducer } from './store';
export { saga } from './saga';
export type { EventsStore } from './store';


type Props = {
  dispatch: any,
  events: Array<Event>,
};
export class InnerEvents extends React.Component<Props> {
  componentDidMount() {
    if (!this.props.events.length) {
      this.props.dispatch(fetchEvents());
    }
  }

  renderList() {
    let rows = this.props.events.map((event, idx) => {
      return <EventRow key={idx} {...event} />;
    });
    return (
      <div className="card-body text-center">
        {rows}
      </div>
    );
  }

  renderNoResults() {
    return (
      <div className="card-body text-center">
        <p className="card-text large-icon"><FontAwesomeIcon icon={fasChartArea} /></p>
        <p className="card-text">No data available</p>
      </div>
    );
  };

  render() {
    const body = this.props.events.length ? this.renderList() : this.renderNoResults();
    return (
      <div className="row">
        <div className="col">
          <div className="card">
            <div className="card-header">Events</div>
            {body}
          </div>
        </div>
      </div>
    );
  };
};


/** Map the redux state to properties passed to component. */
type PartialState = {
  events: EventsStore,
}
export function mapStateToProps(state: PartialState) {
  return {
    events: state.events.events
  };
}


const Events = connect(mapStateToProps, null)(InnerEvents);
export default Events;
