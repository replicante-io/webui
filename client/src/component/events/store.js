'use strict';
//@flow

import { EVENTS_STORE } from './action';

import type { Event } from './action';
import type { EventsAction } from './action';
import type { EventsStoreAction } from './action';


export type EventsStore = {
  events: Array<Event>,
};


export const defaultState: EventsStore = {
  events: [],
};


export function reducer(state: EventsStore = defaultState, action: EventsAction) {
  switch (action.type) {
    case EVENTS_STORE: return storeEvents(state, action);

    default:
      return state;
  }
}


function storeEvents(state: EventsStore, action: EventsStoreAction): EventsStore {
  let newState: EventsStore = {...state};
  newState.events = action.events;
  return newState;
}
