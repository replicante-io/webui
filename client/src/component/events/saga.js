'use strict';
//@flow

import { call, put } from 'redux-saga/effects';
import { takeEvery } from 'redux-saga/effects';

import { EVENTS_FETCH } from './action';
import { EVENTS_STORE } from './action';

import { fetchEvents as fetchEventsApi } from './api';

import type { EventsFetchAction } from './action';


/**
 * Fetches the list of recent events.
 */
export function* fetchEvents(action: EventsFetchAction): any {
  let events = yield call(fetchEventsApi);
  yield put({type: EVENTS_STORE, events: events});
}

/** Main saga for the datafetch component. */
export function* saga(): any {
  yield takeEvery(EVENTS_FETCH, fetchEvents);
}
