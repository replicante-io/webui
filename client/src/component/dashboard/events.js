'use strict';
//@flow

import { Component } from 'react';
import React from 'react';

import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChartArea } from '@fortawesome/free-solid-svg-icons';

import { FetchButton } from '../datafetch';
import { fetchData } from '../datafetch';

import { storeEvents } from '../events/action';
import { fetchEvents as fetchEventsApi } from '../events/api';
import EventRow from '../events/row';
import type { Event } from '../events/action';
import type { EventsStore } from '../events/store';


const FETCH_ID: string = "events-overview";


export function fetchEvents(dispatch: any): () => Promise<void> {
  return () => {
    return fetchEventsApi().then((events) => {
        dispatch(storeEvents(events));
    });
  };
}


type Props = {
  events: Array<Event>,
  dispatch: any,
}

export class InnerEvents extends React.Component<Props> {
  componentDidMount() {
    if (!this.props.events.length) {
      this.props.dispatch(fetchData(FETCH_ID, fetchEvents(this.props.dispatch)));
    }
  }

  renderEvents() {
    let rows = this.props.events.map((event, idx) => {
      return <EventRow key={idx} {...event} />;
    });
    return (
      <div className="card-body text-center">
        <div className="card-text">
          {rows}
        </div>
      </div>
    );
  }

  renderNoData() {
    return (
      <div className="card-body text-center">
        <p className="card-text large-icon"><FontAwesomeIcon icon={faChartArea} /></p>
        <p className="card-text">No data available</p>
      </div>
    );
  }

  render() {
    const body = this.props.events.length ? this.renderEvents() : this.renderNoData();
    return (
      <div className="row space-below">
        <div className="col">
          <div className="card">
            <div className="card-header">
              Latest events
              <FetchButton id={FETCH_ID} request={fetchEvents(this.props.dispatch)} />
            </div>
            {body}
          </div>
        </div>
      </div>
    );
  }
}


/** Map the redux state to properties passed to Button. */
type PartialState = {
  events: EventsStore,
}
export function mapStateToProps(state: PartialState) {
  return {
    events: state.events.events.slice(0, 5),
  };
}


const Events = connect(mapStateToProps, null)(InnerEvents);
export default Events;
