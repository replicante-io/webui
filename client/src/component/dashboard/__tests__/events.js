'use strict';

import '@babel/polyfill';
import fetch from 'jest-fetch-mock';
global.fetch = fetch;

import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import Events from '../events';
import { fetchEvents } from '../events';


const mockStore = configureStore([]);


describe('Dashboard', () => {
  describe('events', () => {

    test('renders with no events', () => {
      let events = [];
      let store = mockStore({
        datafetch: new Map(),
        events: {events: events}
      });
      const tree = renderer.create(
        <Provider store={store}>
          <Events />
        </Provider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('renders with events', () => {
      let events = [{
        data: "SOME DATA",
        event: "AGENT_DOWN",
        timestamp: "2018-04-28T17:57:24.540480643Z"
      }, {
          data: "SOME DATA",
          event: "AGENT_RECOVER",
        timestamp: "2018-04-28T17:50:15.170187929Z"
      }];
      let store = mockStore({
        datafetch: new Map(),
        events: {events: events}
      });
      const tree = renderer.create(
        <Provider store={store}>
          <Events />
        </Provider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    describe('fetchEvents', () => {
      beforeEach(() => {
        fetch.resetMocks()
      });

      test('fulfill', () => {
        let fetcher = fetchEvents(jest.fn());
        fetch.mockResponse(JSON.stringify([{
          data: "SOME DATA",
          event: "AGENT_DOWN",
          timestamp: "2018-04-28T17:57:24.540480643Z"
        }, {
          data: "SOME DATA",
          event: "AGENT_RECOVER",
          timestamp: "2018-04-28T17:50:15.170187929Z"
        }]));
        return fetcher();
      });

      test('reject', () => {
        let error = Error('FAILED');
        let fetcher = fetchEvents(jest.fn());
        fetch.mockReject(error);
        return fetcher().then(
          () => { throw Error('Expected error') },
          (err) => { expect(err).toBe(error) }
        );
      });

      test('reject not 2xx', () => {
        let fetcher = fetchEvents(jest.fn());
        fetch.mockResponse('{}', {status: 500});
        return fetcher().then(
          () => { throw Error('Expected error') },
          () => { }
        );
      });
    });

  });
});
