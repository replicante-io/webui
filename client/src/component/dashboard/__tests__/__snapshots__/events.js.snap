// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard events renders with events 1`] = `
<div
  className="row space-below"
>
  <div
    className="col"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        Latest events
        <button
          className="btn btn-outline-secondary fetch-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-redo-alt fa-w-16 "
            data-icon="redo-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256.455 8c66.269.119 126.437 26.233 170.859 68.685l35.715-35.715C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.75c-30.864-28.899-70.801-44.907-113.23-45.273-92.398-.798-170.283 73.977-169.484 169.442C88.764 348.009 162.184 424 256 424c41.127 0 79.997-14.678 110.629-41.556 4.743-4.161 11.906-3.908 16.368.553l39.662 39.662c4.872 4.872 4.631 12.815-.482 17.433C378.202 479.813 319.926 504 256 504 119.034 504 8.001 392.967 8 256.002 7.999 119.193 119.646 7.755 256.455 8z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
      <div
        className="card-body text-center"
      >
        <div
          className="card-text"
        >
          <div
            className="row event-item"
          >
            <div
              className="col timestamp"
            >
              2018-04-28T17:57:24.540480643Z
            </div>
            <div
              className="col event"
            >
              AGENT_DOWN
            </div>
            <div
              className="col-6 payload codeblock"
            >
              <div
                className="react-json-view"
                style={
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0)",
                    "cursor": "default",
                    "fontFamily": "monospace",
                    "position": "relative",
                  }
                }
              >
                <div
                  className="pretty-json-container object-container"
                >
                  <div
                    className="object-content"
                  >
                    <div
                      className="object-key-val"
                    >
                      <span>
                        <span
                          onClick={[Function]}
                          style={
                            Object {
                              "cursor": "pointer",
                              "display": "inline-block",
                            }
                          }
                        >
                          <div
                            className="icon-container"
                            style={
                              Object {
                                "display": "inline-block",
                                "width": "17px",
                              }
                            }
                          >
                            <span
                              className="collapsed-icon"
                            >
                              <svg
                                fill="currentColor"
                                style={
                                  Object {
                                    "color": "#2aa198",
                                    "height": "1em",
                                    "paddingLeft": "2px",
                                    "verticalAlign": "top",
                                    "width": "1em",
                                  }
                                }
                                viewBox="0 0 15 15"
                              >
                                <path
                                  d="M0 14l6-6-6-6z"
                                />
                              </svg>
                            </span>
                          </div>
                          <span />
                          <span
                            style={
                              Object {
                                "color": "#002b36",
                                "cursor": "pointer",
                                "display": "inline-block",
                                "fontWeight": "bold",
                              }
                            }
                          >
                            {
                          </span>
                        </span>
                      </span>
                      <div
                        className="node-ellipsis"
                        onClick={[Function]}
                        style={
                          Object {
                            "color": "#cb4b16",
                            "cursor": "pointer",
                            "display": "inline-block",
                            "fontSize": "18px",
                            "lineHeight": "10px",
                          }
                        }
                      >
                        ...
                      </div>
                      <span
                        className="brace-row"
                      >
                        <span
                          style={
                            Object {
                              "color": "#002b36",
                              "cursor": "pointer",
                              "display": "inline-block",
                              "fontWeight": "bold",
                              "paddingLeft": "0px",
                            }
                          }
                        >
                          }
                        </span>
                        <div
                          className="object-meta-data"
                          onClick={[Function]}
                          style={
                            Object {
                              "display": "inline-block",
                              "padding": "0px 0px 0px 10px",
                            }
                          }
                        >
                          <span
                            className="object-size"
                            style={
                              Object {
                                "borderRadius": "3px",
                                "color": "rgba(0, 0, 0, 0.3)",
                                "cursor": "default",
                                "fontStyle": "italic",
                                "margin": "0px 6px 0px 0px",
                              }
                            }
                          >
                            1
                             item
                            
                          </span>
                          <span
                            className="copy-to-clipboard-container"
                          >
                            <span
                              onClick={[Function]}
                              style={
                                Object {
                                  "cursor": "pointer",
                                  "display": "inline",
                                }
                              }
                            >
                              <span
                                className="copy-icon"
                              >
                                <svg
                                  fill="currentColor"
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#268bd2",
                                      "fontSize": "15px",
                                      "height": "1em",
                                      "marginRight": "3px",
                                      "verticalAlign": "top",
                                      "width": "1em",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                >
                                  <g>
                                    <path
                                      d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"
                                    />
                                  </g>
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="row event-item"
          >
            <div
              className="col timestamp"
            >
              2018-04-28T17:50:15.170187929Z
            </div>
            <div
              className="col event"
            >
              AGENT_RECOVER
            </div>
            <div
              className="col-6 payload codeblock"
            >
              <div
                className="react-json-view"
                style={
                  Object {
                    "backgroundColor": "rgba(0, 0, 0, 0)",
                    "cursor": "default",
                    "fontFamily": "monospace",
                    "position": "relative",
                  }
                }
              >
                <div
                  className="pretty-json-container object-container"
                >
                  <div
                    className="object-content"
                  >
                    <div
                      className="object-key-val"
                    >
                      <span>
                        <span
                          onClick={[Function]}
                          style={
                            Object {
                              "cursor": "pointer",
                              "display": "inline-block",
                            }
                          }
                        >
                          <div
                            className="icon-container"
                            style={
                              Object {
                                "display": "inline-block",
                                "width": "17px",
                              }
                            }
                          >
                            <span
                              className="collapsed-icon"
                            >
                              <svg
                                fill="currentColor"
                                style={
                                  Object {
                                    "color": "#2aa198",
                                    "height": "1em",
                                    "paddingLeft": "2px",
                                    "verticalAlign": "top",
                                    "width": "1em",
                                  }
                                }
                                viewBox="0 0 15 15"
                              >
                                <path
                                  d="M0 14l6-6-6-6z"
                                />
                              </svg>
                            </span>
                          </div>
                          <span />
                          <span
                            style={
                              Object {
                                "color": "#002b36",
                                "cursor": "pointer",
                                "display": "inline-block",
                                "fontWeight": "bold",
                              }
                            }
                          >
                            {
                          </span>
                        </span>
                      </span>
                      <div
                        className="node-ellipsis"
                        onClick={[Function]}
                        style={
                          Object {
                            "color": "#cb4b16",
                            "cursor": "pointer",
                            "display": "inline-block",
                            "fontSize": "18px",
                            "lineHeight": "10px",
                          }
                        }
                      >
                        ...
                      </div>
                      <span
                        className="brace-row"
                      >
                        <span
                          style={
                            Object {
                              "color": "#002b36",
                              "cursor": "pointer",
                              "display": "inline-block",
                              "fontWeight": "bold",
                              "paddingLeft": "0px",
                            }
                          }
                        >
                          }
                        </span>
                        <div
                          className="object-meta-data"
                          onClick={[Function]}
                          style={
                            Object {
                              "display": "inline-block",
                              "padding": "0px 0px 0px 10px",
                            }
                          }
                        >
                          <span
                            className="object-size"
                            style={
                              Object {
                                "borderRadius": "3px",
                                "color": "rgba(0, 0, 0, 0.3)",
                                "cursor": "default",
                                "fontStyle": "italic",
                                "margin": "0px 6px 0px 0px",
                              }
                            }
                          >
                            1
                             item
                            
                          </span>
                          <span
                            className="copy-to-clipboard-container"
                          >
                            <span
                              onClick={[Function]}
                              style={
                                Object {
                                  "cursor": "pointer",
                                  "display": "inline",
                                }
                              }
                            >
                              <span
                                className="copy-icon"
                              >
                                <svg
                                  fill="currentColor"
                                  preserveAspectRatio="xMidYMid meet"
                                  style={
                                    Object {
                                      "color": "#268bd2",
                                      "fontSize": "15px",
                                      "height": "1em",
                                      "marginRight": "3px",
                                      "verticalAlign": "top",
                                      "width": "1em",
                                    }
                                  }
                                  viewBox="0 0 40 40"
                                >
                                  <g>
                                    <path
                                      d="m30 35h-25v-22.5h25v7.5h2.5v-12.5c0-1.4-1.1-2.5-2.5-2.5h-7.5c0-2.8-2.2-5-5-5s-5 2.2-5 5h-7.5c-1.4 0-2.5 1.1-2.5 2.5v27.5c0 1.4 1.1 2.5 2.5 2.5h25c1.4 0 2.5-1.1 2.5-2.5v-5h-2.5v5z m-20-27.5h2.5s2.5-1.1 2.5-2.5 1.1-2.5 2.5-2.5 2.5 1.1 2.5 2.5 1.3 2.5 2.5 2.5h2.5s2.5 1.1 2.5 2.5h-20c0-1.5 1.1-2.5 2.5-2.5z m-2.5 20h5v-2.5h-5v2.5z m17.5-5v-5l-10 7.5 10 7.5v-5h12.5v-5h-12.5z m-17.5 10h7.5v-2.5h-7.5v2.5z m12.5-17.5h-12.5v2.5h12.5v-2.5z m-7.5 5h-5v2.5h5v-2.5z"
                                    />
                                  </g>
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dashboard events renders with no events 1`] = `
<div
  className="row space-below"
>
  <div
    className="col"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        Latest events
        <button
          className="btn btn-outline-secondary fetch-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-redo-alt fa-w-16 "
            data-icon="redo-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256.455 8c66.269.119 126.437 26.233 170.859 68.685l35.715-35.715C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.75c-30.864-28.899-70.801-44.907-113.23-45.273-92.398-.798-170.283 73.977-169.484 169.442C88.764 348.009 162.184 424 256 424c41.127 0 79.997-14.678 110.629-41.556 4.743-4.161 11.906-3.908 16.368.553l39.662 39.662c4.872 4.872 4.631 12.815-.482 17.433C378.202 479.813 319.926 504 256 504 119.034 504 8.001 392.967 8 256.002 7.999 119.193 119.646 7.755 256.455 8z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
      <div
        className="card-body text-center"
      >
        <p
          className="card-text large-icon"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-chart-area fa-w-16 "
            data-icon="chart-area"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M500 384c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v308h436zM372.7 159.5L288 216l-85.3-113.7c-5.1-6.8-15.5-6.3-19.9 1L96 248v104h384l-89.9-187.8c-3.2-6.5-11.4-8.7-17.4-4.7z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </p>
        <p
          className="card-text"
        >
          No data available
        </p>
      </div>
    </div>
  </div>
</div>
`;
